### DISCUSSION TITLE
Waveshare Pico ResTouch LCD 2.8 SPI ST7789 display

### DISCUSSION THREAD

**Discussion by TimWermer** [peterhinch/micropython-nano-gui#31] [Q&A] [answered]
Hello Peter,

I am trying to get the above display to work with the following code:

[main_2inch8.py.txt](https://github.com/peterhinch/micropython-nano-gui/files/8502377/main_2inch8.py.txt)
[st7789.py.txt](https://github.com/peterhinch/micropython-nano-gui/files/8502378/st7789.py.txt)

I get no errors and my backlight lights up.

Can you give me a hint how to proceed?

Best regards,

Tim Wermer

**Comment by peterhinch** [comment]
It is hard for me to support hardware I don't possess, however this display should work. The Waveshare link says it comes with a MicroPython demo. Does that work?

Have you tried the procedure [here](https://github.com/peterhinch/micropython-nano-gui/blob/master/README.md#23-verifying-hardware-configuration)? This is the best way to verify your hardware and your `color_setup.py` file. Also the SPI bus will appear to work even if not connected - please double check your wiring.

**Reply by TimWermer** [reply]
![Pico](https://user-images.githubusercontent.com/6475207/164791733-71467c6e-d53a-463c-8dac-c619a5806a80.jpg)

I am using a Pimoroni double deck for the pico. So i am pretty confident the wiring is ok. 
I had a problem rshell was hanging when i was using your instructions. I found an earlier post by you with the following command that did the trick. 

`rshell -p /dev/ttyACM0  --buffer-size=30`

The problem i am facing now is that in color_setup.py i cannot figure out why the following code breaks:

`pdc = machine.Pin('Y1', machine.Pin.OUT_PP, value=0)
pcs = machine.Pin('Y2', machine.Pin.OUT_PP, value=1)
prst = machine.Pin('Y3', machine.Pin.OUT_PP, value=1)`

Can you tell me where the attribute OUT_PP is defined?

Thanks again for your help!

**Reply by peterhinch** [reply]
The constant `OUT_PP` is Pyboard-specific. For other targets it should be `OUT`.

**Reply by TimWermer** [reply]
Ok thanks Peter!
i changed the color_setup.py into the following :

`import machine
import gc

from drivers.st7789.st7789_4bit import ST7789 as SSD

height = 320  # 2.8 inch Pico ResTouch 320*240 (rows*cols) display
width = 240  


pdc = machine.Pin(8 , machine.Pin.OUT, value=0)
pcs = machine.Pin(9, machine.Pin.OUT, value=1)
prst = machine.Pin(15, machine.Pin.OUT, value=1)
spi = machine.SPI(1, 30_000_000, sck=machine.Pin(10), mosi=machine.Pin(11), miso=machine.Pin(12))
gc.collect()  # Precaution before instantiating framebuf
ssd = SSD(spi, pcs, pdc, prst, height, width)  # Create a display instance`

If i run aclock.py i get no errors but no screen what so ever.

>>> %Run -c $EDITOR_CONTENT
Orientation: Horizontal. Reversal: False. Width: 240. Height: 320.
Start row = 0 col = 0

Do you see anything strange in the above code?
![Pico-ResTouch-LCD-2 8-details-inter](https://user-images.githubusercontent.com/6475207/164945256-5bf84d27-4bbf-4f50-8a78-9ceafac4027e.jpg)

**Comment by peterhinch** [comment]
My Pico [example](https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/st7789_pico.py) has MISO as GPIO 8, which is the pin identified on the Pico diagram. I don't know if this matters as there is no data coming back, but it might be worth a try.

Also look at [this example](https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/st7789_pico_lcd_114.py) which is for a Waveshare display. Note that some special args were required. It's quite likely you'll need `display=TDISPLAY`. Unfortunately these display controller chips can be wired up in different ways: supporting the tdisplay took quite a lot of work. Going "off piste" with displays that aren't specifically tested can be challenging.

Good luck.

**Comment by peterhinch** [comment]
If you get it to run please post your setup file and I will publish it (with your attribution).

**Reply by TimWermer** [reply]
Hello Peter, i am trying to get some more info from Waveshare. They are using your code as drivers for the device. I will certainly inform you about any progress i will make!

**Reply by TimWermer** [reply]
So....remember you advised to check wiring? And i replied it is fixed on a Pimorone Pico Omnibus? Well after talking to Waveshare they asked me to click the Pico directly on the Pico ResTouch. I did so and guess what... it works. 
This means that The Pico ResTouch 2.8 can be added to your compatible devices list!
Peter thank you for your patience and help.

![IMG_5438](https://user-images.githubusercontent.com/6475207/166059111-b94cbb21-86a3-4008-a971-5e7f43161f07.jpg)

**Comment by peterhinch** [comment]
Please could you post your setup file and I'll publish it. Thank you.

**Comment by TimWermer** [comment]
Hello Peter,

Here are the files as requested.

Thanks again for your great work!

Tim
[main_2inch8.py.txt](https://github.com/peterhinch/micropython-nano-gui/files/8599537/main_2inch8.py.txt)
[st7789.py.txt](https://github.com/peterhinch/micropython-nano-gui/files/8599538/st7789.py.txt)

**Comment by peterhinch** [comment]
I hadn't spotted the changes you made to the device driver. I notice your driver initialises `lut` but doesn't use it. The 16-bit driver would need further adaptation to remove this if it were to be compatible with either of my GUI's. I didn't produce a 16-bit driver because of the huge buffer size. I won't publish the driver as I think the buffer is too large especially for the larger `microgui` codebase. 

I've [posted](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#332-waveshare-pico-restouch-lcd) your setup adapted to use the 4-bit driver.

**Reply by TimWermer** [reply]
Using the 4-bit driver results in the attached display:
![IMG_5487](https://user-images.githubusercontent.com/6475207/166524708-4be1f2cf-50f9-4ef2-91f3-0ba399ecf12a.jpg)
There are no errors reported.

**Comment by peterhinch** [comment]
OK, I can't see why the 4-bit driver isn't working. Given that your 16-bit version works I'm sure it can be fixed. I'll order one of these displays and (hopefully) post a solution.

**Reply by TimWermer** [reply]
Hello Peter, i can send you one if you like? More then fair for the work you do for the Community!
Just let me know. tim@wermer.org.

**Reply by peterhinch** [reply]
Thank you for the offer, Tim, but I've already ordered one. I'll post a comment once I have the 4-bit driver working.

**Comment by peterhinch** [comment, marked as answer]
I received a board today. [This setup file](https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/ws_pico_res_touch.py) enables nano-gui to run with the 4-bit driver. It can be run in landscape or portrait mode, but the SSD constructor args to achieve these modes are rather unexpected. See code comments.

**Reply by TimWermer** [reply]
Hello Peter,

That works like a charm! 

![IMG_5503](https://user-images.githubusercontent.com/6475207/167272402-f8240b8a-452d-471e-9473-2ae459f1625a.jpg)

For the Waveshare demo i get the following error:

AttributeError: 'ST7789' object has no attribute 'wait'

What library/repo do you use for the SD-card?

Fantastic job! 

Have a good evening.

Tim

**Comment by peterhinch** [comment]
I'm glad you have it working. The file `waveshare_test.py` is specifically for ePaper displays and will not run on other types. I should make it produce a more friendly error message.

The file you need is the official driver [sdcard.py](https://github.com/micropython/micropython/blob/master/drivers/sdcard/sdcard.py).

I like the board except that it has no easy way to access unused GPIO pins. To run `micro-gui` I'll need to solder wires onto the Pico for the pushbuttons. Most users will want to connect something to the board.

**Reply by TimWermer** [reply]
I have all GPIO's available as you can see in the picture below.

![IMG_5510](https://user-images.githubusercontent.com/6475207/167312038-043d40b4-fd33-4122-84d8-e77bfee642a4.png)

I used the following product from Kiwi Electronis:

https://www.kiwi-electronics.nl/en/36-pin-stacking-header-5-pack-2875?search=pins&page=3

Maybe that helps.
